cudax/include/cuda/experimental/__execution/env.cuh

```diff
@@ -0,0 +1,108 @@
+//===----------------------------------------------------------------------===//
+//
+// Part of CUDA Experimental in CUDA C++ Core Libraries,
+// under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+// SPDX-FileCopyrightText: Copyright (c) 2024 NVIDIA CORPORATION & AFFILIATES.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef __CUDAX___EXECUTION_ENV_CUH
+#define __CUDAX___EXECUTION_ENV_CUH
+
+#include <cuda/std/detail/__config>
+
+#if defined(_CCCL_IMPLICIT_SYSTEM_HEADER_GCC)
+#  pragma GCC system_header
+#elif defined(_CCCL_IMPLICIT_SYSTEM_HEADER_CLANG)
+#  pragma clang system_header
+#elif defined(_CCCL_IMPLICIT_SYSTEM_HEADER_MSVC)
+#  pragma system_header
+#endif // no system header
+
+#include <cuda/__memory_resource/properties.h>
+#include <cuda/std/__type_traits/is_same.h>
+#include <cuda/std/__utility/forward.h>
+#include <cuda/std/__utility/move.h>
+
+#include <cuda/experimental/__async/queries.cuh>
+#include <cuda/experimental/__execution/policy.cuh>
+#include <cuda/experimental/__memory_resource/any_resource.cuh>
+#include <cuda/experimental/__memory_resource/device_memory_resource.cuh>
+#include <cuda/experimental/__memory_resource/get_memory_resource.cuh>
+#include <cuda/experimental/__stream/get_stream.cuh>
+#include <cuda/experimental/__stream/stream_ref.cuh>
+
+namespace cuda::experimental::execution
+{
+
+template <class... _Properties>
+class env_t
+{
+private:
+  using __resource   = ::cuda::experimental::mr::any_resource<_Properties...>;
+  using __stream_ref = ::cuda::experimental::stream_ref;
+
+  __resource __mr_           = ::cuda::experimental::mr::device_memory_resource{};
+  __stream_ref __stream_     = cudaStream_t{0};
+  execution_policy __policy_ = execution_policy::unsequenced_device;
+
+public:
+  //! @brief Default constructs an environment using
+  //! * ``::cuda::experimental::mr::device_memory_resource`` as the resource
+  //! * the default stream
+  //! * ``execution_policy::unsequenced_device`` as the execution policy
+  _CCCL_HIDE_FROM_ABI env_t() = default;
+
+  //! @brief Construct an env_t from an any_resource, a stream and a policy
+  //! @param __mr The any_resource passed in
+  //! @param __stream The stream_ref passed in
+  //! @param __policy The execution_policy passed in
+  _CCCL_HIDE_FROM_ABI env_t(__resource __mr,
```

Comment:
will we ever want to build an environment in device code?

Reply:
i know. that's why i'm asking. so? will we ever want to build an environment in device code?

Reply:
Right now I would say no, but we can revise that later